File: harness/cbor.ts
export type CborScalar = null | boolean | number | string | Uint8Array;
export type CborValue =
  | CborScalar
  | CborValue[]
  | { [key: string]: CborValue }
  | Map<CborScalar, CborValue>;

const utf8 = new TextEncoder();
const strictUtf8 = new TextDecoder("utf-8", { fatal: true });

function concat(parts: readonly Uint8Array[]): Uint8Array {
  const length = parts.reduce((sum, part) => sum + part.length, 0);
  const out = new Uint8Array(length);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

function head(major: number, value: number): Uint8Array {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new TypeError(`CBOR integer/length is not a non-negative safe integer: ${value}`);
  }
  if (value < 24) return Uint8Array.of((major << 5) | value);
  if (value <= 0xff) return Uint8Array.of((major << 5) | 24, value);
  if (value <= 0xffff) {
    return Uint8Array.of((major << 5) | 25, value >>> 8, value);
  }
  if (value <= 0xffff_ffff) {
    return Uint8Array.of(
      (major << 5) | 26,
      value / 0x1_00_00_00_00,
      value >>> 16,
      value >>> 8,
      value,
    );
  }
  const high = Math.floor(value / 0x1_00_00_00_00);
  const low = value - high * 0x1_00_00_00_00;
  return Uint8Array.of(
    (major << 5) | 27,
    high >>> 24,
    high >>> 16,
    high >>> 8,
    high,
    low >>> 24,
    low >>> 16,
    low >>> 8,
    low,
  );
}

function compareBytes(left: Uint8Array, right: Uint8Array): number {
  if (left.length !== right.length) return left.length - right.length;
  for (let i = 0; i < left.length; i += 1) {
    const difference = left[i]! - right[i]!;
    if (difference !== 0) return difference;
  }
  return 0;
}

function encodeMap(entries: readonly (readonly [CborScalar, CborValue])[]): Uint8Array {
  const encoded = entries.map(([key, value]) => ({
    key: encodeCbor(key),
    value: encodeCbor(value),
  }));
  encoded.sort((a, b) => compareBytes(a.key, b.key));
  for (let i = 1; i < encoded.length; i += 1) {
    if (compareBytes(encoded[i - 1]!.key, encoded[i]!.key) === 0) {
      throw new TypeError("duplicate CBOR map key");
    }
  }
  return concat([
    head(5, encoded.length),
    ...encoded.flatMap(({ key, value }) => [key, value]),
  ]);
}

export function encodeCbor(value: CborValue): Uint8Array {
  if (value === null) return Uint8Array.of(0xf6);
  if (value === false) return Uint8Array.of(0xf4);
  if (value === true) return Uint8Array.of(0xf5);
  if (typeof value === "number") {
    if (!Number.isSafeInteger(value)) {
      throw new TypeError("CBOR profile permits only safe integers, not floats");
    }
    return value >= 0 ? head(0, value) : head(1, -1 - value);
  }
  if (typeof value === "string") {
    const bytes = utf8.encode(value);
    return concat([head(3, bytes.length), bytes]);
  }
  if (value instanceof Uint8Array) {
    return concat([head(2, value.length), value]);
  }
  if (Array.isArray(value)) {
    return concat([head(4, value.length), ...value.map(encodeCbor)]);
  }
  if (value instanceof Map) {
    return encodeMap([...value.entries()] as [CborScalar, CborValue][]);
  }
  if (typeof value === "object") {
    return encodeMap(
      Object.entries(value).map(([key, item]) => [key, item] as const),
    );
  }
  throw new TypeError(`unsupported CBOR value: ${String(value)}`);
}

export interface DecodeOptions {
  strict?: boolean;
  maxDepth?: number;
}

class Decoder {
  private offset = 0;

  constructor(
    private readonly bytes: Uint8Array,
    private readonly maxDepth: number,
  ) {}

  decode(): CborValue {
    const value = this.item(0);
    if (this.offset !== this.bytes.length) throw new Error("trailing CBOR bytes");
    return value;
  }

  private take(): number {
    const byte = this.bytes[this.offset];
    if (byte === undefined) throw new Error("truncated CBOR item");
    this.offset += 1;
    return byte;
  }

  private argument(additional: number): number {
    if (additional < 24) return additional;
    const count = additional === 24 ? 1 : additional === 25 ? 2 : additional === 26 ? 4 : additional === 27 ? 8 : 0;
    if (count === 0) {
      if (additional === 31) throw new Error("indefinite-length CBOR is forbidden");
      throw new Error("reserved CBOR additional information");
    }
    let value = 0;
    for (let i = 0; i < count; i += 1) value = value * 256 + this.take();
    if (!Number.isSafeInteger(value)) throw new Error("CBOR integer exceeds safe range");
    return value;
  }

  private bytesOf(length: number): Uint8Array {
    const end = this.offset + length;
    if (end > this.bytes.length) throw new Error("truncated CBOR string");
    const value = this.bytes.slice(this.offset, end);
    this.offset = end;
    return value;
  }

  private item(depth: number): CborValue {
    if (depth > this.maxDepth) throw new Error("CBOR nesting exceeds configured maximum");
    const initial = this.take();
    const major = initial >>> 5;
    const additional = initial & 31;
    if (major === 6) throw new Error("CBOR tags are forbidden");
    if (major === 7) {
      if (additional === 20) return false;
      if (additional === 21) return true;
      if (additional === 22) return null;
      if (additional === 25 || additional === 26 || additional === 27) {
        throw new Error("CBOR floats are forbidden");
      }
      throw new Error("unsupported CBOR simple value");
    }
    const argument = this.argument(additional);
    if (major === 0) return argument;
    if (major === 1) return -1 - argument;
    if (major === 2) return this.bytesOf(argument);
    if (major === 3) {
      try {
        return strictUtf8.decode(this.bytesOf(argument));
      } catch {
        throw new Error("invalid UTF-8 in CBOR text string");
      }
    }
    if (major === 4) {
      return Array.from({ length: argument }, () => this.item(depth + 1));
    }
    if (major === 5) {
      const entries: [CborScalar, CborValue][] = [];
      const keys = new Set<string>();
      let stringsOnly = true;
      for (let i = 0; i < argument; i += 1) {
        const key = this.item(depth + 1);
        if (!isScalar(key)) throw new Error("compound CBOR map keys are outside the AAR profile");
        const keyId = toHex(encodeCbor(key));
        if (keys.has(keyId)) throw new Error("duplicate CBOR map key");
        keys.add(keyId);
        stringsOnly &&= typeof key === "string";
        entries.push([key, this.item(depth + 1)]);
      }
      if (!stringsOnly) return new Map(entries);
      const object: Record<string, CborValue> = {};
      for (const [key, value] of entries) {
        Object.defineProperty(object, key as string, {
          value,
          enumerable: true,
          configurable: true,
          writable: true,
        });
      }
      return object;
    }
    throw new Error(`unsupported CBOR major type ${major}`);
  }
}

function isScalar(value: CborValue): value is CborScalar {
  return value === null || typeof value === "boolean" || typeof value === "number" || typeof value === "string" || value instanceof Uint8Array;
}

export function decodeCbor(bytes: Uint8Array, options: DecodeOptions = {}): CborValue {
  const decoded = new Decoder(bytes, options.maxDepth ?? 32).decode();
  if (options.strict) {
    const canonical = encodeCbor(decoded);
    if (!equalBytes(bytes, canonical)) throw new Error("non-canonical CBOR encoding");
  }
  return decoded;
}

export function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  return left.length === right.length && left.every((byte, index) => byte === right[index]);
}

export function toHex(bytes: Uint8Array): string {
  return [...bytes].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

export function fromHex(hex: string): Uint8Array {
  if (!/^(?:[0-9a-fA-F]{2})*$/.test(hex)) throw new TypeError("invalid hex");
  return Uint8Array.from(hex.match(/../g)?.map((pair) => Number.parseInt(pair, 16)) ?? []);
}

File: pyref/cbor.py
"""Minimal RFC 8949 core deterministic CBOR codec used by the C1 KATs."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


class CBORError(ValueError):
    """Raised for CBOR outside the AAR deterministic encoding profile."""

    def __init__(self, message: str, code: str = "cbor/malformed") -> None:
        super().__init__(message)
        self.code = code


def _head(major: int, value: int) -> bytes:
    if value < 0:
        raise CBORError("negative CBOR argument")
    prefix = major << 5
    if value < 24:
        return bytes((prefix | value,))
    if value <= 0xFF:
        return bytes((prefix | 24, value))
    if value <= 0xFFFF:
        return bytes((prefix | 25,)) + value.to_bytes(2, "big")
    if value <= 0xFFFFFFFF:
        return bytes((prefix | 26,)) + value.to_bytes(4, "big")
    if value <= 0xFFFFFFFFFFFFFFFF:
        return bytes((prefix | 27,)) + value.to_bytes(8, "big")
    raise CBORError("integer exceeds CBOR uint64")


def dumps(value: Any) -> bytes:
    """Encode an AAR value with RFC 8949 section 4.2.1 ordering."""
    if value is None:
        return b"\xf6"
    if value is False:
        return b"\xf4"
    if value is True:
        return b"\xf5"
    if isinstance(value, int):
        return _head(0, value) if value >= 0 else _head(1, -1 - value)
    if isinstance(value, bytes):
        return _head(2, len(value)) + value
    if isinstance(value, str):
        encoded = value.encode("utf-8")
        return _head(3, len(encoded)) + encoded
    if isinstance(value, (list, tuple)):
        return _head(4, len(value)) + b"".join(dumps(item) for item in value)
    if isinstance(value, dict):
        entries = [(dumps(key), dumps(item)) for key, item in value.items()]
        entries.sort(key=lambda entry: (len(entry[0]), entry[0]))
        return _head(5, len(entries)) + b"".join(
            key_bytes + item_bytes for key_bytes, item_bytes in entries
        )
    raise CBORError(f"unsupported CBOR type: {type(value).__name__}")


@dataclass
class _Decoder:
    data: bytes
    offset: int = 0
    max_depth: int = 32

    def _take(self, size: int) -> bytes:
        end = self.offset + size
        if end > len(self.data):
            raise CBORError("truncated CBOR item", "cbor/malformed")
        result = self.data[self.offset:end]
        self.offset = end
        return result

    def _argument(self, additional: int) -> int:
        if additional < 24:
            return additional
        sizes = {24: 1, 25: 2, 26: 4, 27: 8}
        size = sizes.get(additional)
        if size is None:
            if additional == 31:
                raise CBORError(
                    "indefinite-length CBOR is forbidden", "cbor/indefinite-length"
                )
            raise CBORError("reserved CBOR additional information")
        value = int.from_bytes(self._take(size), "big")
        if (size == 1 and value < 24) or (
            size == 2 and value <= 0xFF
        ) or (size == 4 and value <= 0xFFFF) or (
            size == 8 and value <= 0xFFFFFFFF
        ):
            raise CBORError("non-shortest CBOR argument", "cbor/non-canonical")
        return value

    def item(self, depth: int = 0) -> Any:
        initial = self._take(1)[0]
        major, additional = initial >> 5, initial & 0x1F
        if major == 0:
            return self._argument(additional)
        if major == 1:
            return -1 - self._argument(additional)
        if major == 2:
            return self._take(self._argument(additional))
        if major == 3:
            raw = self._take(self._argument(additional))
            try:
                return raw.decode("utf-8")
            except UnicodeDecodeError as exc:
                raise CBORError("invalid UTF-8 text string", "cbor/invalid-utf8") from exc
        if major == 4:
            if depth >= self.max_depth:
                raise CBORError("CBOR nesting depth exceeds 32", "resource/cbor-depth")
            return [self.item(depth + 1) for _ in range(self._argument(additional))]
        if major == 5:
            if depth >= self.max_depth:
                raise CBORError("CBOR nesting depth exceeds 32", "resource/cbor-depth")
            count = self._argument(additional)
            result: dict[Any, Any] = {}
            encoded_keys: set[bytes] = set()
            previous_order: tuple[int, bytes] | None = None
            for _ in range(count):
                key_start = self.offset
                key = self.item(depth + 1)
                key_bytes = self.data[key_start:self.offset]
                if key_bytes in encoded_keys:
                    raise CBORError("duplicate map key", "cbor/duplicate-key")
                encoded_keys.add(key_bytes)
                order = (len(key_bytes), key_bytes)
                if previous_order is not None and order <= previous_order:
                    raise CBORError(
                        "non-deterministic map-key ordering", "cbor/non-canonical"
                    )
                previous_order = order
                try:
                    result[key] = self.item(depth + 1)
                except TypeError as exc:
                    raise CBORError("unhashable map key") from exc
            return result
        if major == 6:
            raise CBORError("CBOR tags are forbidden", "cbor/tag-forbidden")
        if additional == 20:
            return False
        if additional == 21:
            return True
        if additional == 22:
            return None
        if additional in {25, 26, 27}:
            raise CBORError("CBOR floating-point values are forbidden", "cbor/float-forbidden")
        raise CBORError("unsupported CBOR simple value")


def loads(data: bytes, *, max_depth: int = 32) -> Any:
    """Decode exactly one deterministic AAR-profile CBOR item."""
    decoder = _Decoder(data, max_depth=max_depth)
    value = decoder.item()
    if decoder.offset != len(data):
        raise CBORError("trailing bytes after CBOR item", "cbor/trailing-bytes")
    return value
